src/emmy/tape.cljc

```diff
@@ -0,0 +1,706 @@
+#_"SPDX-License-Identifier: GPL-3.0";;
+
+(ns emmy.tape
+  "This namespace contains an implementation of [[TapeCell]], a type that forms
+  the basis for the reverse-mode automatic differentiation implementation in
+  Emmy."
+  (:require [emmy.differential :as d]
+            [emmy.function :as f]
+            [emmy.generic :as g]
+            [emmy.matrix :as matrix]
+            [emmy.structure :as s]
+            [emmy.util :as u]
+            [emmy.value :as v]))
+
+;; ## Reverse-mode Automatic Differentiation
+;;
+;; > The AD implementation developed in this namespace is based on Alexey
+;; > Radul's [implementation in
+;; > dvl](https://github.com/axch/dysvunctional-language).
+;;
+;; This namespace develops an implementation of "reverse-mode" automatic
+;; differentation, in contrast with the forward mode AD implementation
+;; in [[emmy.differential]]. These two are closely related, and the
+;; implementations could merge once I get around to reading and
+;; implementing [the YOLO paper](https://arxiv.org/abs/2204.10923).
+;;
+;; The core idea of forward-mode AD is that we can use the chain rule to
+;; mechanically build up a partial derivative by wrapping one of the dependent
+;; variables in a "dual number" and accumulating the derivative as the
+;; calculation progresses.
+;;
+;; For functions with one input and many outputs, this style computes the full
+;; partial derivative in one pass.
+;;
+;; The downside of forward-mode AD is that each partial derivative requires
+;; running the full computation for each independent input variable. For
+;; functions with many inputs and a single output (like the Langrangian), this
+;; linear scaling is not great.
+
+;; "Reverse-mode" AD is an alternative way of building up all of the partial
+;; derivatives for such a function in a single pass.
+;;
+;; > NOTE I plan on fleshing out a full, worked example of the how the chain
+;; > rule works in reverse-mode... but please accept this sketch for now.
+;;
+;; The approach works like this:
+;;
+;; First, wrap all inputs to the function in an instance of a type called
+;; a [[TapeCell]]. This type holds
+;;
+;;   - a "tag", used as in [[emmy.differential]] to prevent confusion
+;;     between [[TapeCell]] instances created for different nested runs of
+;;     differentation
+;;   - a unique ID, used to de-duplicate work in the reverse pass
+;;   - the "primal value" wrapped by this [[TapeCell]] instance
+;;   - A map of the [[TapeCell]]'s inputs => the partial derivative of the
+;;     primal value with respect to that input
+;;
+;; Pass the augmented inputs into the original function, which should be built
+;; out of primitives that are overloaded to deal with [[TapeCell]] instances.
+;;
+;; A primitive `f` of `n` arguments, given [[TapeCell]] inputs, should return a
+;; new [[TapeCell]] of the form:
+;;
+;; ```clojure
+;; {:tag <input tag>
+;;  :id <fresh ID>
+;;  :primal (f a b c ...)
+;;  :in->partials {<tape-a> <(((partial 0) f) a b c ...)>
```

Comment:
`<tape-a>` etc. are references to other tape cells, pointing "in" to this one? Could they store the partials themselves I wonder